routing/localchans/manager_test.go

```diff
@@ -91,13 +96,23 @@ func TestManager(t *testing.T) {
 	forAllOutgoingChannels := func(cb func(*channeldb.ChannelEdgeInfo,
 		*channeldb.ChannelEdgePolicy) error) error {
 
-		return cb(
+		err := cb(
 			&channeldb.ChannelEdgeInfo{
 				Capacity:     chanCap,
 				ChannelPoint: chanPoint,
 			},
 			&currentPolicy,
 		)
+		if err != nil {
+			return err
+		}
+		return cb(
+			&channeldb.ChannelEdgeInfo{
+				Capacity:     chanCapInactive,
+				ChannelPoint: chanPointInactive,
+			},
+			&currentPolicy,
+		)
```

Comment:
Not happy with this bit. Is there a more elegant solution?

Reply:
How about something like this rather?

```
func TestManager(t *testing.T) {
	t.Parallel()
	
	var (
		chanPointValid    = wire.OutPoint{Hash: chainhash.Hash{1}, Index: 2}
		chanCap           = btcutil.Amount(1000)
		chanPointInactive = wire.OutPoint{Hash: chainhash.Hash{2}, Index: 2}
		chanPointMissing  = wire.OutPoint{Hash: chainhash.Hash{3}, Index: 2}
		maxPendingAmount  = lnwire.MilliSatoshi(999000)
		minHTLC           = lnwire.MilliSatoshi(2000)
	)

	newPolicy := routing.ChannelPolicy{
		FeeSchema: routing.FeeSchema{
			BaseFee: 100,
			FeeRate: 200,
		},
		TimeLockDelta: 80,
		MaxHTLC:       5000,
	}

	currentPolicy := channeldb.ChannelEdgePolicy{
		MinHTLC:      minHTLC,
		MessageFlags: lnwire.ChanUpdateOptionMaxHtlc,
	}

	updateForwardingPolicies := func(
		chanPolicies map[wire.OutPoint]htlcswitch.ForwardingPolicy) {

		if len(chanPolicies) == 0 {
			return
		}

		if len(chanPolicies) != 1 {
			t.Fatal("unexpected number of policies to apply")
		}

		policy := chanPolicies[chanPointValid]
		if policy.TimeLockDelta != newPolicy.TimeLockDelta {
			t.Fatal("unexpected time lock delta")
		}
		if policy.BaseFee != newPolicy.BaseFee {
			t.Fatal("unexpected base fee")
		}
		if uint32(policy.FeeRate) != newPolicy.FeeRate {
			t.Fatal("unexpected base fee")
		}
		if policy.MaxHTLC != newPolicy.MaxHTLC {
			t.Fatal("unexpected max htlc")
		}
	}

	var expectedNumUpdates int
	propagateChanPolicyUpdate := func(
		edgesToUpdate []discovery.EdgeWithInfo) error {

		if len(edgesToUpdate) != expectedNumUpdates {
			t.Fatalf("unexpected number of updates,"+
				" expected %d got %d", expectedNumUpdates,
				len(edgesToUpdate))
		}

		for _, edge := range edgesToUpdate {
			policy := edge.Edge
			if !policy.MessageFlags.HasMaxHtlc() {
				t.Fatal("expected max htlc flag")
			}
			if policy.TimeLockDelta != uint16(newPolicy.TimeLockDelta) {
				t.Fatal("unexpected time lock delta")
			}
			if policy.FeeBaseMSat != newPolicy.BaseFee {
				t.Fatal("unexpected base fee")
			}
			if uint32(policy.FeeProportionalMillionths) != newPolicy.FeeRate {
				t.Fatal("unexpected base fee")
			}
			if policy.MaxHTLC != newPolicy.MaxHTLC {
				t.Fatal("unexpected max htlc")
			}
		}

		return nil
	}

	type channel struct {
		active   bool
		edgeInfo *channeldb.ChannelEdgeInfo
	}

	var channelSet []channel

	forAllOutgoingChannels := func(cb func(*channeldb.ChannelEdgeInfo,
		*channeldb.ChannelEdgePolicy) error) error {

		for _, c := range channelSet {
			if err := cb(c.edgeInfo, &currentPolicy); err != nil {
				return err
			}
		}
		return nil
	}

	fetchChannel := func(chanPoint wire.OutPoint) (*channeldb.OpenChannel,
		error) {

		constraints := channeldb.ChannelConstraints{
			MaxPendingAmount: maxPendingAmount,
			MinHTLC:          minHTLC,
		}

		return &channeldb.OpenChannel{
			LocalChanCfg: channeldb.ChannelConfig{
				ChannelConstraints: constraints,
			},
		}, nil
	}

	hasActiveLink := func(chanID lnwire.ChannelID) bool {
		for _, c := range channelSet {
			if !chanID.IsChanPoint(&c.edgeInfo.ChannelPoint) {
				continue
			}

			return c.active
		}
		return false
	}

	manager := Manager{
		UpdateForwardingPolicies:  updateForwardingPolicies,
		PropagateChanPolicyUpdate: propagateChanPolicyUpdate,
		ForAllOutgoingChannels:    forAllOutgoingChannels,
		FetchChannel:              fetchChannel,
		HasActiveLink:             hasActiveLink,
	}

	// Policy with no max htlc value.
	MaxHTLCPolicy := currentPolicy
	MaxHTLCPolicy.MaxHTLC = newPolicy.MaxHTLC
	noMaxHtlcPolicy := newPolicy
	noMaxHtlcPolicy.MaxHTLC = 0

	tests := []struct {
		name                string
		currentPolicy       channeldb.ChannelEdgePolicy
		newPolicy           routing.ChannelPolicy
		channelSet          []channel
		specifiedChanPoints []wire.OutPoint
		expectedNumUpdates  int
		expectErr           bool
	}{
		{
			name:          "valid channel",
			currentPolicy: currentPolicy,
			newPolicy:     newPolicy,
			channelSet: []channel{
				{
					edgeInfo: &channeldb.ChannelEdgeInfo{
						Capacity:     chanCap,
						ChannelPoint: chanPointValid,
					},
					active: true,
				},
			},
			specifiedChanPoints: []wire.OutPoint{chanPointValid},
			expectedNumUpdates:  1,
		},
		{
			name:          "one active and one inactive channel",
			currentPolicy: currentPolicy,
			newPolicy:     newPolicy,
			channelSet: []channel{
				{
					edgeInfo: &channeldb.ChannelEdgeInfo{
						Capacity:     chanCap,
						ChannelPoint: chanPointValid,
					},
					active: true,
				},
				{
					edgeInfo: &channeldb.ChannelEdgeInfo{
						Capacity:     chanCap,
						ChannelPoint: chanPointInactive,
					},
					active: false,
				},
			},
			specifiedChanPoints: []wire.OutPoint{
				chanPointValid,
				chanPointInactive,
			},
			expectErr:          true,
			expectedNumUpdates: 1,
		},
		{
			name:          "all channels",
			currentPolicy: currentPolicy,
			newPolicy:     newPolicy,
			channelSet: []channel{
				{
					edgeInfo: &channeldb.ChannelEdgeInfo{
						Capacity:     chanCap,
						ChannelPoint: chanPointValid,
					},
					active: true,
				},
			},
			specifiedChanPoints: []wire.OutPoint{},
			expectedNumUpdates:  1,
		},
		{
			name:          "missing channel",
			currentPolicy: currentPolicy,
			newPolicy:     newPolicy,
			channelSet: []channel{
				{
					edgeInfo: &channeldb.ChannelEdgeInfo{
						Capacity:     chanCap,
						ChannelPoint: chanPointValid,
					},
					active: true,
				},
			},
			specifiedChanPoints: []wire.OutPoint{chanPointMissing},
			expectErr:           true,
			expectedNumUpdates:  0,
		},
		{
			// Here, no max htlc is specified, the max htlc value
			// should be kept unchanged.
			name:          "no mac htlc specified",
			currentPolicy: MaxHTLCPolicy,
			newPolicy:     noMaxHtlcPolicy,
			channelSet: []channel{
				{
					edgeInfo: &channeldb.ChannelEdgeInfo{
						Capacity:     chanCap,
						ChannelPoint: chanPointValid,
					},
					active: true,
				},
			},
			specifiedChanPoints: []wire.OutPoint{chanPointValid},
			expectedNumUpdates:  1,
		},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			currentPolicy = test.currentPolicy
			channelSet = test.channelSet
			expectedNumUpdates = test.expectedNumUpdates

			err := manager.UpdatePolicy(test.newPolicy, test.specifiedChanPoints...)
			if err != nil && test.expectErr {
				return
			}
			if err != nil {
				t.Fatalf(err.Error())
			}
			if test.expectErr {
				t.Fatalf("expected error")
			}
		})
	}
}
```

Reply:
@ellemouton Is there a reason you removed the testing code you posted two days ago? I did copy it and are running tests right now.

Reply:
@ellemouton It is better than what I initially wrote btw. Impressed.

Reply:
oops sorry! for some reason it looked like i had commented it twice so i removed one 🙈  
will post again for reference:
```
func TestManager(t *testing.T) {
	t.Parallel()
	
	var (
		chanPointValid    = wire.OutPoint{Hash: chainhash.Hash{1}, Index: 2}
		chanCap           = btcutil.Amount(1000)
		chanPointInactive = wire.OutPoint{Hash: chainhash.Hash{2}, Index: 2}
		chanPointMissing  = wire.OutPoint{Hash: chainhash.Hash{3}, Index: 2}
		maxPendingAmount  = lnwire.MilliSatoshi(999000)
		minHTLC           = lnwire.MilliSatoshi(2000)
	)

	newPolicy := routing.ChannelPolicy{
		FeeSchema: routing.FeeSchema{
			BaseFee: 100,
			FeeRate: 200,
		},
		TimeLockDelta: 80,
		MaxHTLC:       5000,
	}

	currentPolicy := channeldb.ChannelEdgePolicy{
		MinHTLC:      minHTLC,
		MessageFlags: lnwire.ChanUpdateOptionMaxHtlc,
	}

	updateForwardingPolicies := func(
		chanPolicies map[wire.OutPoint]htlcswitch.ForwardingPolicy) {

		if len(chanPolicies) == 0 {
			return
		}

		if len(chanPolicies) != 1 {
			t.Fatal("unexpected number of policies to apply")
		}

		policy := chanPolicies[chanPointValid]
		if policy.TimeLockDelta != newPolicy.TimeLockDelta {
			t.Fatal("unexpected time lock delta")
		}
		if policy.BaseFee != newPolicy.BaseFee {
			t.Fatal("unexpected base fee")
		}
		if uint32(policy.FeeRate) != newPolicy.FeeRate {
			t.Fatal("unexpected base fee")
		}
		if policy.MaxHTLC != newPolicy.MaxHTLC {
			t.Fatal("unexpected max htlc")
		}
	}

	var expectedNumUpdates int
	propagateChanPolicyUpdate := func(
		edgesToUpdate []discovery.EdgeWithInfo) error {

		if len(edgesToUpdate) != expectedNumUpdates {
			t.Fatalf("unexpected number of updates,"+
				" expected %d got %d", expectedNumUpdates,
				len(edgesToUpdate))
		}

		for _, edge := range edgesToUpdate {
			policy := edge.Edge
			if !policy.MessageFlags.HasMaxHtlc() {
				t.Fatal("expected max htlc flag")
			}
			if policy.TimeLockDelta != uint16(newPolicy.TimeLockDelta) {
				t.Fatal("unexpected time lock delta")
			}
			if policy.FeeBaseMSat != newPolicy.BaseFee {
				t.Fatal("unexpected base fee")
			}
			if uint32(policy.FeeProportionalMillionths) != newPolicy.FeeRate {
				t.Fatal("unexpected base fee")
			}
			if policy.MaxHTLC != newPolicy.MaxHTLC {
				t.Fatal("unexpected max htlc")
			}
		}

		return nil
	}

	type channel struct {
		active   bool
		edgeInfo *channeldb.ChannelEdgeInfo
	}

	var channelSet []channel

	forAllOutgoingChannels := func(cb func(*channeldb.ChannelEdgeInfo,
		*channeldb.ChannelEdgePolicy) error) error {

		for _, c := range channelSet {
			if err := cb(c.edgeInfo, &currentPolicy); err != nil {
				return err
			}
		}
		return nil
	}

	fetchChannel := func(chanPoint wire.OutPoint) (*channeldb.OpenChannel,
		error) {

		constraints := channeldb.ChannelConstraints{
			MaxPendingAmount: maxPendingAmount,
			MinHTLC:          minHTLC,
		}

		return &channeldb.OpenChannel{
			LocalChanCfg: channeldb.ChannelConfig{
				ChannelConstraints: constraints,
			},
		}, nil
	}

	hasActiveLink := func(chanID lnwire.ChannelID) bool {
		for _, c := range channelSet {
			if !chanID.IsChanPoint(&c.edgeInfo.ChannelPoint) {
				continue
			}

			return c.active
		}
		return false
	}

	manager := Manager{
		UpdateForwardingPolicies:  updateForwardingPolicies,
		PropagateChanPolicyUpdate: propagateChanPolicyUpdate,
		ForAllOutgoingChannels:    forAllOutgoingChannels,
		FetchChannel:              fetchChannel,
		HasActiveLink:             hasActiveLink,
	}

	// Policy with no max htlc value.
	MaxHTLCPolicy := currentPolicy
	MaxHTLCPolicy.MaxHTLC = newPolicy.MaxHTLC
	noMaxHtlcPolicy := newPolicy
	noMaxHtlcPolicy.MaxHTLC = 0

	tests := []struct {
		name                string
		currentPolicy       channeldb.ChannelEdgePolicy
		newPolicy           routing.ChannelPolicy
		channelSet          []channel
		specifiedChanPoints []wire.OutPoint
		expectedNumUpdates  int
		expectErr           bool
	}{
		{
			name:          "valid channel",
			currentPolicy: currentPolicy,
			newPolicy:     newPolicy,
			channelSet: []channel{
				{
					edgeInfo: &channeldb.ChannelEdgeInfo{
						Capacity:     chanCap,
						ChannelPoint: chanPointValid,
					},
					active: true,
				},
			},
			specifiedChanPoints: []wire.OutPoint{chanPointValid},
			expectedNumUpdates:  1,
		},
		{
			name:          "one active and one inactive channel",
			currentPolicy: currentPolicy,
			newPolicy:     newPolicy,
			channelSet: []channel{
				{
					edgeInfo: &channeldb.ChannelEdgeInfo{
						Capacity:     chanCap,
						ChannelPoint: chanPointValid,
					},
					active: true,
				},
				{
					edgeInfo: &channeldb.ChannelEdgeInfo{
						Capacity:     chanCap,
						ChannelPoint: chanPointInactive,
					},
					active: false,
				},
			},
			specifiedChanPoints: []wire.OutPoint{
				chanPointValid,
				chanPointInactive,
			},
			expectErr:          true,
			expectedNumUpdates: 1,
		},
		{
			name:          "all channels",
			currentPolicy: currentPolicy,
			newPolicy:     newPolicy,
			channelSet: []channel{
				{
					edgeInfo: &channeldb.ChannelEdgeInfo{
						Capacity:     chanCap,
						ChannelPoint: chanPointValid,
					},
					active: true,
				},
			},
			specifiedChanPoints: []wire.OutPoint{},
			expectedNumUpdates:  1,
		},
		{
			name:          "missing channel",
			currentPolicy: currentPolicy,
			newPolicy:     newPolicy,
			channelSet: []channel{
				{
					edgeInfo: &channeldb.ChannelEdgeInfo{
						Capacity:     chanCap,
						ChannelPoint: chanPointValid,
					},
					active: true,
				},
			},
			specifiedChanPoints: []wire.OutPoint{chanPointMissing},
			expectErr:           true,
			expectedNumUpdates:  0,
		},
		{
			// Here, no max htlc is specified, the max htlc value
			// should be kept unchanged.
			name:          "no mac htlc specified",
			currentPolicy: MaxHTLCPolicy,
			newPolicy:     noMaxHtlcPolicy,
			channelSet: []channel{
				{
					edgeInfo: &channeldb.ChannelEdgeInfo{
						Capacity:     chanCap,
						ChannelPoint: chanPointValid,
					},
					active: true,
				},
			},
			specifiedChanPoints: []wire.OutPoint{chanPointValid},
			expectedNumUpdates:  1,
		},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			currentPolicy = test.currentPolicy
			channelSet = test.channelSet
			expectedNumUpdates = test.expectedNumUpdates

			err := manager.UpdatePolicy(test.newPolicy, test.specifiedChanPoints...)
			if err != nil && test.expectErr {
				return
			}
			if err != nil {
				t.Fatalf(err.Error())
			}
			if test.expectErr {
				t.Fatalf("expected error")
			}
		})
	}
}
```